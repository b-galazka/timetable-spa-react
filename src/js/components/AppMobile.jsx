import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import AppMobileLoader from './AppMobileLoader';
import { getMobileAppData } from '../actions/mobileAppData';

import texts from '../../json/texts.json';

import '../../scss/mobileApp.scss';

function mapStateToProps(state) {

    const { data, fetched } = state.mobileAppData;

    return {
        mobileAppData: data,
        mobileAppDataFetched: fetched
    };
}

function mapDisptachToProps(dispatch) {

    return {
        getMobileAppData: bindActionCreators(getMobileAppData, dispatch)
    };
}

class AppMobile extends Component {

    render() {

        const { mobileAppData, mobileAppDataFetched } = this.props;

        return (
            <main className="mobile-app">

                {
                    (mobileAppDataFetched) ?

                    <div className="mobile-app__wrapper">
                        <a
                            className="mobile-app__button button"
                            href={mobileAppData.apkFileUrl}
                            target="_blank"
                        >
                            {texts.downloadApp}
                        </a>

                        <button
                            className="mobile-app__button button"
                            onClick={this.props.visitPage}
                        >
                            {texts.visitPage}
                        </button>
                    </div> :

                    <AppMobileLoader />
                }
                
            </main>
        );
    }

    componentDidMount() {

        this.props.getMobileAppData();
    }
}

export default connect(mapStateToProps, mapDisptachToProps)(AppMobile);