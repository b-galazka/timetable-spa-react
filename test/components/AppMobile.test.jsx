import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import renderer from 'react-test-renderer';

import AppMobile from '../../src/js/components/AppMobile';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../src/js/components/AppMobileLoader', () => 'AppMobileLoader');

jest.mock(
    '../../src/json/texts',
    () => ({ downloadApp: 'download app', visitPage: 'visit page' })
);

describe('AppMobile component', () => {

    let appMobile;

    beforeEach(() => {

        appMobile = shallow(
            <AppMobile.WrappedComponent
                visitPage={jest.fn()}
                mobileAppDataFetched={false}
                getMobileAppData={jest.fn()}
            />
        ).instance();

        appMobile.props = { ...appMobile.props };
    });

    it('should render loader if mobile app data has not been fetched yet', () => {

        const tree = renderer.create(
            <AppMobile.WrappedComponent
                visitPage={jest.fn()}
                mobileAppDataFetched={false}
                getMobileAppData={jest.fn()}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render page if mobile app data has been already fetched', () => {

        const tree = renderer.create(
            <AppMobile.WrappedComponent
                visitPage={jest.fn()}
                mobileAppDataFetched
                getMobileAppData={jest.fn()}
                mobileAppData={{ apkFileUrl: 'apk file url' }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should call props.visitPage method on button click', () => {

        const visitPageMock = jest.fn();

        const appMobile = mount(
            <AppMobile.WrappedComponent
                visitPage={visitPageMock}
                mobileAppDataFetched
                getMobileAppData={jest.fn()}
                mobileAppData={{ apkFileUrl: 'apk file url' }}
            />
        );

        appMobile.find('button.mobile-app__button').simulate('click');

        expect(visitPageMock).toHaveBeenCalledTimes(1);
    });

    describe('AppMobile.prototype.componentDidMount', () => {

        it('should call props.getMobileAppData method', () => {

            const getMobileAppDataMock = jest.fn();

            appMobile.props.getMobileAppData = getMobileAppDataMock;

            appMobile.componentDidMount();

            expect(getMobileAppDataMock).toHaveBeenCalledTimes(1);
        });
    });
});