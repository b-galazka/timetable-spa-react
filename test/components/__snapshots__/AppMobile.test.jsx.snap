// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppMobile component should render loader if mobile app data has not been fetched yet 1`] = `
<main
  className="mobile-app"
>
  <AppMobileLoader />
</main>
`;

exports[`AppMobile component should render page if mobile app data has been already fetched 1`] = `
<main
  className="mobile-app"
>
  <div
    className="mobile-app__wrapper"
  >
    <a
      className="mobile-app__button button"
      href="apk file url"
      rel="noopener noreferrer"
      target="_blank"
    >
      download app
    </a>
    <button
      className="mobile-app__button button"
      onClick={[MockFunction]}
    >
      visit page
    </button>
  </div>
</main>
`;
